From 48d06f24e83c740d1a5f0d463497f265c5fa0d02 Mon Sep 17 00:00:00 2001
From: Gaurao Chaudhari <quic_gaurchau@quicinc.com>
Date: Tue, 29 Mar 2022 10:52:32 -0700
Subject: [PATCH] net: skbuff_recycle: Fix the skb consume condition in
 recycler

Instead of using the skb pointers to determine the size of skb
during its consumption in the recycler, we should use the initially
allocated memory size to determine the length of the skb and eventually
consuming it based on that size.

Change-Id: I851ab7c9ab11b2813693fd3df6f595c8abe2b9eb
Signed-off-by: Gaurao Chaudhari <quic_gaurchau@quicinc.com>
---
 net/core/skbuff_recycle.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/net/core/skbuff_recycle.h
+++ b/net/core/skbuff_recycle.h
@@ -140,11 +140,11 @@ static inline bool consume_skb_can_recyc
 		return false;
 
 	min_skb_size = SKB_DATA_ALIGN(min_skb_size + NET_SKB_PAD);
-	if (unlikely(skb_end_pointer(skb) - skb->head < min_skb_size))
+	if (unlikely(SKB_WITH_OVERHEAD(ksize(skb->head)) < min_skb_size))
 		return false;
 
 	max_skb_size = SKB_DATA_ALIGN(max_skb_size + NET_SKB_PAD);
-	if (unlikely(skb_end_pointer(skb) - skb->head > max_skb_size))
+	if (unlikely(SKB_WITH_OVERHEAD(ksize(skb->head)) > max_skb_size))
 		return false;
 
 	if (unlikely(skb_cloned(skb)))
